### DISCUSSION TITLE
Preserving State with Link Component and searchParams After Page Refresh

### DISCUSSION THREAD

**Discussion by komzweb** [vercel/next.js#65341] [Help] [answered]
### Summary

I'm implementing a filter using `searchParams` and `Link` component. This code basically works fine. But, when I refresh the browser with the query parameter `/?status=deleted` and then click the home `/` link, the deleted post is rendered. Using `useEffect` and `router.refresh` solves the problem for now, but is there any other solution?

### Additional information

```js
export default async function Home({
  searchParams,
}: {
  searchParams: { status: 'active' | 'deleted' | undefined }
}) {
  const status = searchParams.status

  let posts: FetchedPostListItem[] = []

  if (status === 'deleted') {
    posts = await getDeletedPosts()
  } else {
    posts = await getPosts()
  }

  return (
    <div>
      <Link href="/">Home</Link>
      <Link href="/?status=active">Active</Link>
      <Link href="/?status=deleted">Deleted</Link>
      <PostList posts={posts} />
    </div>
  )
}
```

**Comment by solo-samurai** [comment]
@komzweb This code works as expected and shows deleted posts only when the status is deleted.

There is no need to refresh the router. what's your intended behavior?
 
```ts

import Link from "next/link";

interface Post {
  id: number;
  title: string;
  content: string;
}

async function getPosts() {
  return [
    {
      id: 1,
      title: "Post 1",
      content: "Content 1",
    },
    {
      id: 2,
      title: "Post 2",
      content: "Content 2",
    },
    {
      id: 3,
      title: "Post 3",
      content: "Content 3",
    },
  ];
}

async function getDeletedPosts() {
  return [
    {
      id: 1,
      title: "Deleted Post 1",
      content: "Content 1",
    },
    {
      id: 2,
      title: "Deleted Post 2",
      content: "Content 2",
    },
    {
      id: 3,
      title: "Deleted Post 3",
      content: "Content 3",
    },
  ];
}

export default async function Home({
  searchParams,
}: {
  searchParams: { status: "active" | "deleted" | undefined };
}) {
  const status = searchParams.status;

  let posts: Post[] = [];

  if (status === "deleted") {
    posts = await getDeletedPosts();
  } else {
    posts = await getPosts();
  }

  return (
    <div>
      <Link href="/">Home</Link>
      <Link href="/?status=active">Active</Link>
      <Link href="/?status=deleted">Deleted</Link>
      <PostList posts={posts} />
    </div>
  );
}

function PostList({ posts }: { posts: Post[] }) {
  return (
    <div>
      {posts.map((post: any) => (
        <div key={post.id}>
          <h1>{post.title}</h1>
          <p>{post.content}</p>
        </div>
      ))}
    </div>
  );
}
```

**Reply by komzweb** [reply]
First, click the `Deleted` link to render the list of deleted posts and refresh your browser. Then navigate to `/` using the `Home` link.
The query parameters have been removed from the URL in the search bar, but the list of deleted posts is still rendered on the page. This is the problem.

**Reply by solo-samurai** [reply]
I followed your steps and ran the code above, it works as expected. 
maybe there is more to your code or project to reproduce the issue. someone might have faced your issue and could help.

**Reply by komzweb** [reply]
To test this issue, I created a new project with the latest Next.js (14.2.3). Then, I copied and pasted the code you created and ran it, but it still didn't work as expected. It doesn't seem to be a problem with my own project or the version of Next.js. I tried it with Chrome, Safari, Firefox, and other browsers, so it doesn't seem to be a browser problem either. I don't think I can solve it right away, so I'll wait and see until I get some hints. Thanks for your help.

**Reply by solo-samurai** [reply]
@komzweb you're correct this issue happens on the build server. 

the reason next.js router cache https://nextjs.org/docs/app/building-your-application/caching#opting-out-3

and you can't opt out. for now at least.

however instead `<Link href="/">Home</Link>`  for home page use `<Link href="/?status=active">Home</Link>` 
otherwise, you will get the previous cache for / `/?status=active` or `/?status=deleted`

#### searchParams change will cause the router cache to update.

**Reply by komzweb** [reply]
You're right. Setting the parameter in the Home link is probably the best solution for now. I can finally move forward.

**Comment by komzweb** [comment, marked as answer]
Setting `<Link href="/" prefetch={true}>` and then starting in production solved the problem. I completely forgot that `prefetch` doesn't work in development.

Since the issue is resolved, I'll close this discussion.
